packages/back-end/src/queryRunners/QueryRunner.ts

```diff
@@ -191,13 +213,77 @@ export abstract class QueryRunner<
     });
   }
 
+  public async startReadyQueries(): Promise<void> {
+    const queuedQueries = await getQueriesByIds(
+      this.model.organization,
+      this.model.queries
+        .filter((q) => q.status === "queued")
+        .map((q) => q.query)
+    );
+    logger.debug(
+      `Starting any queued queries that are ready: ${queuedQueries.map(
+        (q) => q.id
+      )}`
+    );
+    queuedQueries.map(async (query) => {
+      // check if all dependencies are finished
+      // assumes all dependencies are within the model; if any are not, query will hang
+      // in queued state
+
+      const failedDependencies: QueryPointer[] = [];
+      const succeededDependencies: QueryPointer[] = [];
+      const pendingDependencies: QueryPointer[] = [];
+
+      const dependencyIds: string[] = query.dependencies ?? [];
+      dependencyIds.forEach((dependencyId) => {
+        const dependencyQuery = this.model.queries.find(
+          (q) => q.query == dependencyId
+        );
+        if (dependencyQuery === undefined) {
+          throw new Error(`Dependency ${dependencyId} not found in model`);
+        } else if (dependencyQuery.status === "succeeded") {
+          succeededDependencies.push(dependencyQuery);
+        } else if (dependencyQuery.status === "failed") {
+          failedDependencies.push(dependencyQuery);
+        } else {
+          pendingDependencies.push(dependencyQuery);
+        }
+      });
+
+      if (failedDependencies.length) {
+        logger.debug(`${query.id}: Dependency failed...`);
+        updateQuery(query, {
+          finishedAt: new Date(),
+          status: "failed",
+          error: `Dependencies failed: ${failedDependencies.map(
+            (q) => q.query
+          )}`,
+        });
+        return;
+      }
+      if (pendingDependencies.length) {
+        logger.debug(`${query.id}: Dependencies pending...`);
+        return;
+      }
+      if (succeededDependencies.length === dependencyIds.length) {
+        logger.debug(`${query.id}: Dependencies completed, running...`);
+        const { run, process } = this.runCallbacks[query.id];
+        this.executeQuery(query, run, process);
+      }
+    });
+  }
+
   public async refreshQueryStatuses(): Promise<void> {
     const oldStatus = this.getOverallQueryStatus();
     logger.debug("Refreshing query statuses");
 
-    // If there are no running queries, return immediately
-    if (!this.model.queries.some((q) => q.status === "running")) {
-      logger.debug("No running queries, return");
+    // If there are no running or queued queries, return immediately
+    if (
+      !this.model.queries.some(
+        (q) => q.status === "running" || q.status === "queued"
+      )
+    ) {
+      logger.debug("No running or queued queries, return");
```

Comment:
I don't think we need this change if we do the above tweaks (awaiting the update calls and calling `onQueryFinish()` at the end)

Reply:
I think you're right, but I didn't make this change because we now rely on getting the queryMap for `startReadyQueries` and getting that first before this call changes the point of this check, I think?